examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py

```diff
@@ -0,0 +1,609 @@
+    logger.info(f"  Total optimization steps = {args.max_train_steps}")
+    # Only show the progress bar once on each machine.
+    progress_bar = tqdm(range(args.max_train_steps), disable=not accelerator.is_local_main_process)
+    completed_steps = 0
+
+    for epoch in range(args.num_train_epochs):
+        model.train()
+        for step, batch in enumerate(train_dataloader):
+            outputs = model(**batch)
+            loss = outputs.loss
+            loss = loss / args.gradient_accumulation_steps
+            accelerator.backward(loss)
+            if step % args.gradient_accumulation_steps == 0 or step == len(train_dataloader) - 1:
+                optimizer.step()
+                lr_scheduler.step()
+                optimizer.zero_grad()
+                progress_bar.update(1)
+                completed_steps += 1
+
+            if completed_steps >= args.max_train_steps:
+                break
+
+            # Log all results
+            if (step + 1) % (args.gradient_accumulation_steps * args.logging_steps) == 0:
+                loss.detach()
+
+                if accelerator.state.num_processes > 1:
+                    loss = accelerator.gather(loss).sum() / accelerator.num_processes
+
+                train_logs = {
+                    "loss": loss,
+                    "lr": torch.tensor(optimizer.param_groups[0]["lr"]),
+                }
+
+                # Evaluate (gather required)
+                with torch.no_grad():
+                    upsampled_logits = torch.nn.functional.interpolate(
+                        outputs.logits, size=batch["labels"].shape[-2:], mode="bilinear", align_corners=False
+                    )
+                    predictions = upsampled_logits.argmax(dim=1)
+
+                    metric.add_batch(
+                        predictions=accelerator.gather(predictions),
+                        references=accelerator.gather(batch["labels"]),
+                    )
+                train_metrics = metric.compute(
+                    num_labels=len(id2label),
+                    ignore_index=255,
+                    reduce_labels=False,  # we've already reduced the labels before
+                )
+                train_logs["mean_iou"] = train_metrics["mean_iou"]
+                train_logs["mean_accuracy"] = train_metrics["mean_accuracy"]
+                train_logs["overall_accuracy"] = train_metrics["overall_accuracy"]
+
+                log_str = ""
+                for k, v in train_logs.items():
+                    if isinstance(v, torch.Tensor):
+                        log_str += "| {}: {:.3e}".format(k, v.item())
+                    else:
+                        log_str += "| {}: {:.3e}".format(k, v)
+
+                if accelerator.is_local_main_process:
+                    progress_bar.write(log_str)
+                    if is_wandb_available() and args.wandb:
+                        wandb.log(train_logs)
```

Comment:
This can be refactored some based on your preference. I recommend something like this:
```suggestion
                if accelerator.is_local_main_process:
                    progress_bar.write(log_str)
                if args.wandb:
                    accelerator.log(train_logs)
```